ledger/testing/randomAccounts_test.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package testing
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/test/partitiontest"
+	"github.com/algorand/go-algorand/test/reflectionhelpers"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestAccounts(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	accountDataType := reflect.TypeOf(basics.AccountData{})
+
+	referencedAccountTypes := make([]reflectionhelpers.TypePath, 0)
+	reflectionhelpers.IterateReferencedTypes(accountDataType, func(path reflectionhelpers.TypePath, stack []reflect.Type) bool {
+		if len(path) == 0 {
+			return true
+		}
+		if path[len(path)-1].FieldName == "_struct" && stack[len(stack)-1] == reflect.TypeOf(struct{}{}) {
+			return true
+		}
+		referencedAccountTypes = append(referencedAccountTypes, path.Clone())
+		return true
+	})
+
+	// If this test becomes flaky, increase niter
+	niter := 1000
+
+	accountFieldSeenZero := make([]bool, len(referencedAccountTypes))
+	accountFieldSeenNonzero := make([]bool, len(referencedAccountTypes))
+
+	accounts := RandomAccounts(niter, false)
+	for _, account := range accounts {
+		accountValue := reflect.ValueOf(account)
+		for i, typePath := range referencedAccountTypes {
+			values := typePath.ResolveValues(accountValue)
+
+			for _, value := range values {
+				isZero := value.IsZero()
+				if value.Kind() == reflect.Slice || value.Kind() == reflect.Map {
+					fieldLen := value.Len()
+					isZero = fieldLen == 0
+				}
+				if !accountFieldSeenZero[i] && isZero {
+					accountFieldSeenZero[i] = true
+				}
+				if !accountFieldSeenNonzero[i] && !isZero {
+					accountFieldSeenNonzero[i] = true
+				}
+			}
+		}
+	}
+
+	// It's ok for these fields to never be the zero value
```

Comment:
@jasonpaulos What is the guiding principle(s) for knowing when it's safe to extend the _zero value_ exception list?

I'm wondering if there's intuition that can be added as an inline comment.  I have a _guess_, but it feels fuzzy to me.

Reply:
Missed a key phrase above:  to extend the _never zero value_ list*

Reply:
I think the intuition here is that it would be invalid to write an account with a zero value for that field to our database. Depending on the field, that either means that the field is non-optional, or that the zero value of the field's type is an unachievable/invalid value.

If that alights with your thoughts, I can add it as an inline comment.

Reply:
Added in d6f6199fd